### PR TITLE
[WIP] Support router serialization.

### PR DESCRIPTION
**Pull request by nathanhammond**
Currently works for all scenarios excepting late-additions via `delegate`. This may support those and I may just have broken tests, but I'm unsure.

Current status:
- [x] Make it work.
- [ ] Make it right.
- [ ] Make it fast.

### CODE REVIEW

File: tests/router-tests.js [hunk 1]
```diff
@@ -13,7 +13,12 @@ function resultsMatch(results, array, queryParams) {
 
 module("The match DSL", {
   setup: function() {
-    router = new RouteRecognizer();
+    router.map = function() {
+      var original = new RouteRecognizer();
+      original.map(...arguments);
```
Comment (chadhietala) [inline tests/router-tests.js]: This probably the parse error.

**Comment by nathanhammond** [comment]
Uh-oh, benchmarks aren't good:
```
testing
- Add
- End-to-end
- Generate
- Handlers For
- Recognize
- Serialized
running first test, please wait...
  Add ................... 505.99 op/s
  End-to-end ............ 178.20 op/s
  Generate ........ 4,081,694.96 op/s
  Handlers For ... 32,213,961.87 op/s
  Recognize ......... 322,362.30 op/s
  Serialized .............. 0.69 op/s
```

I'm not sure why it's slow yet.

File: bench/benches/serialize.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+var serialized = router.toJSON();
+
+module.exports = {
+  name: 'Serialized',
+  fn: function() {
+    new RouteRecognizer(serialized);
+    
+    // Look up time is constant
+    router.recognize('/posts/1');
```
Comment (chadhietala) [inline bench/benches/serialize.js]: This is looking the wrong RR instance.

**Comment by nathanhammond** [comment]
@krisselden Can I steal some of your time to learn what I'm doing wrong on the performance side here? I'm going to delay aiming for 100% compatibility until I know it makes sense to do so.

File: bench/benches/serialize.js (hunk 2 above)
Comment (nathanhammond) [inline bench/benches/serialize.js]: Good catch, rerunning benchmark now.

Comment (nathanhammond) [inline bench/benches/serialize.js]: No faster: it wouldn't be, thinking about it. That `new RouteRecognizer(serialized);` is stupidly expensive for some reason.

**Comment by nathanhammond** [comment]
Found the problem:
```bash
$ ls -la out.json 
-rw-r--r--  1 user  group  63414305 May  3 18:44 out.json
```

`JSON.parse` simply can't handle a 60mb JSON object. Also, can you say memory pressure? This needs to be optimized (and I think it can be, pretty dramatically).

**Comment by nathanhammond** [comment]
Can we start having the conversation about whether/how to land the dramatic refactoring currently living in trie.js once I finish making it 100% API compatible?

**Comment by nathanhammond** [comment]
> 166 assertions of 178 passed, 12 failed.

Closing in on correct, taking a look at early benchmarks. Add and End-to-end seem to be doing something terribly wrong; I suspect that the `compact` method needs some work. The recognize benchmark is effectively a worst-case-scenario as the generated flat list doesn't lend itself to prefix trie optimizations.

We're likely going to trade runtime performance for serialized performance here (in add), but I want to keep runtime within striking distance of serialized. This means that Add and End-to-end need to reach hundreds of op/s.

I need to rebuild serialization for the trie method to see what we're looking at for rehydration speed as well as JSON file size.

---

| Benchmark | Original | Trie |
| --- | --- | --- |
| Add | 489.09 op/s | Crash |
| End-to-end | 229.27 op/s | 0.54 op/s |
| Generate | 4,032,441.12 op/s | 1,467,045.61 op/s |
| Handlers For | 29,003,809.27 op/s | 18,566,106.55 op/s |
| Recognize | 352,315.79 op/s | 25,253.27 op/s |
| Serialize | 0.69 op/s | |

**Comment by nathanhammond** [comment]
New benchmarks with JIT compacting. JIT `compact`ing spreads out the cost in `recognize` and `generate` making the first call to each super-slow. I'm guessing, based upon the non-existent benchmark change, that it's throwing out that run as an outlier. 😜

| Benchmark | Original | Trie | JIT Trie |
| --- | --- | --- | --- |
| Add | 489.09 op/s | Crash |  167.12 op/s |
| End-to-end | 229.27 op/s | 0.54 op/s | 0.59 op/s |
| Generate | 4,032,441.12 op/s | 1,467,045.61 op/s | 1,617,403.23 op/s |
| Handlers For | 29,003,809.27 op/s | 18,566,106.55 op/s | 17,657,366.47 op/s |
| Recognize | 352,315.79 op/s | 25,253.27 op/s | 25,979.61 op/s |
| Serialize | 0.69 op/s | | |

**Comment by nathanhammond** [comment]
Current status:
- Rebased on top of @bantic's work. 339 out of 355 tests passing. I know of additional correctness issues that I need to revisit with regards to specificity that we don't have tests for.
- Many tests now run after a serialization/deserialization pass to check for idempotence of that operation.
- I've not re-reviewed performance recently.
- I need to clean up some of the mess I've made in integrating @bantic's work.

**Comment by nathanhammond** [comment]
Superseded by [ember-route-recognizer](https://github.com/nathanhammond/ember-route-recognizer).

**Pull request closed without merging**